### PR TITLE
ENH add logistic regression

### PR DESCRIPTION
**Pull request by QB3**

### CODE REVIEW

File: andersoncd/datafits.py [hunk 1]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
     def gradient_scalar(self, X, w, Xw, j):
         return (X[:, j] @ Xw - self.Xty[j]) / len(Xw)
 
-    def gradient_scalar_sparse(self, Xj, idx_nz, Xw, j):
+    def gradient_scalar_sparse(self, Xj, y, idx_nz, Xw, j):
```
Comment (mathurinm) [inline andersoncd/datafits.py:68]: if you add it here you need it everywhere no ?

Comment (mathurinm) [reply andersoncd/datafits.py:68]: otherwise we make it self.y ? wdyt ?

File: andersoncd/datafits.py [hunk 2]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
...
         XjTXw = 0
         for i, idx_i in enumerate(idx_nz):
             XjTXw += Xj[i] * Xw[idx_i]
         return (XjTXw - self.Xty[j]) / len(Xw)
+
+
+from numba import njit
```
Comment (mathurinm) [inline andersoncd/datafits.py:75]: top level

File: andersoncd/datafits.py [hunk 3]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
...
+
+
+@jitclass(spec_quadratic)
+class Logistic(BaseDatafit):
+    def __init__(self):
+        pass
+
+    def initialize(self, X, y):
+        self.Xty = X.T @ y
+        self.lipschitz = (X ** 2).sum(axis=0) / len(y) / 4
```
Comment (mathurinm) [inline andersoncd/datafits.py:89]: do you prefer a / b / c or a / (b * c) ?

File: andersoncd/datafits.py [hunk 4]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
...
+    return 1. / (1. + np.exp(- x))
+
+
+@jitclass(spec_quadratic)
+class Logistic(BaseDatafit):
+    def __init__(self):
+        pass
+
+    def initialize(self, X, y):
+        self.Xty = X.T @ y
```
Comment (mathurinm) [inline andersoncd/datafits.py:88]: not needed, right ?

File: andersoncd/datafits.py [hunk 5]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
...
+        self.lipschitz = np.empty(n_features)
+        for j in range(n_features):
+            Xj = X_data[X_indptr[j]:X_indptr[j+1]]
+            idx_nz = X_indices[X_indptr[j]:X_indptr[j+1]]
+            for i, idx_i in enumerate(idx_nz):
+                self.Xty[j] += Xj[i] * y[idx_i]
+            self.lipschitz[j] = (Xj ** 2).sum() / len(y) / 4
+
+    def value(self, y, w, Xw):
+        return - np.log(sigmoid(y * Xw)).sum() / len(y)
```
Comment (mathurinm) [inline andersoncd/datafits.py:105]: My 2 cents : I'd write it log(1 + np.exp(-y  * Xw))   that I find more readable

File: andersoncd/datafits.py [hunk 6]
```diff
@@ -65,8 +65,52 @@ def value(self, y, w, Xw):
...
+
+    def gradient_scalar(self, X, y, w, Xw, j):
+        return (X[:, j] @ (y * (sigmoid(y * Xw) - 1))) / len(y)
```
Comment (mathurinm) [inline andersoncd/datafits.py:108]: you can avoid the -1, no ?

**Comment by QB3** [comment]
Remaining:
- [ ] gradient for sparse matrices
- [ ] remove y in call for grads

File: andersoncd/estimators.py [hunk 7]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+    --------
+    celer_path
+
+    References
+    ----------
+    .. [1] Q. Bertrand, M. Massias
+      "Anderson acceleration of coordinate descent", AISTATS 2021,
+      http://proceedings.mlr.press/v130/bertrand21a.html
+
+    .. [2] M. Massias, A. Gramfort, J. Salmon
```
Comment (mathurinm) [inline andersoncd/estimators.py:545]: rm

File: andersoncd/estimators.py [hunk 8]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+    >>> from andersoncd import LogisticRegression
+    >>> clf = LogisticRegression(C=1.)
+    >>> clf.fit([[0, 0], [1, 1], [2, 2]], [0, 1, 1])
+    LogisticRegression(C=1.0, penalty='l1', tol=0.0001, fit_intercept=False,
+    max_iter=50, verbose=False, max_epochs=50000, p0=10, warm_start=False)
+
+    >>> print(clf.coef_)
+    [[0.4001237  0.01949392]]
+
+    See also
```
Comment (mathurinm) [inline andersoncd/estimators.py:535]: rm

File: andersoncd/estimators.py [hunk 9]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+
+        `coef_` is of shape (1, n_features) when the given problem is binary.
+
+    intercept_ :  ndarray of shape (1,) or (n_classes,)
+        constant term in decision function. Not handled yet.
+
+    n_iter_ : int
+        Number of subproblems solved by Celer to reach the specified tolerance.
+
+    Examples
```
Comment (mathurinm) [inline andersoncd/estimators.py:524]: rm

File: andersoncd/estimators.py [hunk 10]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+    coef_ : ndarray of shape (1, n_features) or (n_classes, n_features)
+        Coefficient of the features in the decision function.
+
+        `coef_` is of shape (1, n_features) when the given problem is binary.
+
+    intercept_ :  ndarray of shape (1,) or (n_classes,)
+        constant term in decision function. Not handled yet.
+
+    n_iter_ : int
+        Number of subproblems solved by Celer to reach the specified tolerance.
```
Comment (mathurinm) [inline andersoncd/estimators.py:522]: fix

File: andersoncd/estimators.py [hunk 11]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+    max_epochs : int
+        Maximum number of CD epochs on each subproblem.
+
+    p0 : int
+        First working set size.
+
+    warm_start : bool, optional (default=False)
+        When set to True, reuse the solution of the previous call to fit as
+        initialization, otherwise, just erase the previous solution.
+        Only False is supported so far.
```
Comment (mathurinm) [inline andersoncd/estimators.py:504]: true ?

File: andersoncd/estimators.py [hunk 12]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+
+    penalty : 'l1'.
+        Other penalties are not supported.
+
+    solver : "andersoncd"
+        Other solvers are not supported.
+
+    tol : float, optional
+        Stopping criterion for the optimization: the solver runs until the
+        duality gap is smaller than ``tol * len(y) * log(2)`` or the
```
Comment (mathurinm) [inline andersoncd/estimators.py:483]: true ?

File: andersoncd/estimators.py [hunk 13]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+
+    Parameters
+    ----------
+    C : float, default=1.0
+        Inverse of regularization strength; must be a positive float.
+
+    penalty : 'l1'.
+        Other penalties are not supported.
+
+    solver : "andersoncd"
```
Comment (mathurinm) [inline andersoncd/estimators.py:478]: rm solver no ?

File: andersoncd/estimators.py [hunk 14]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+
+class LogisticRegression(LogReg_sklearn):
+    """
+    Sparse Logistic regression scikit-learn estimator based on Celer solver.
+
+    The optimization objective for sparse Logistic regression is::
+
+    mean(log(1 + e^{-y_i x_i^T w})) + 1. / C * ||w||_1
+
+    The solvers use a working set strategy. To solve problems restricted to a
```
Comment (mathurinm) [inline andersoncd/estimators.py:467]: merge sentences

File: andersoncd/estimators.py [hunk 15]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
             prune=self.prune, verbose=self.verbose)
+
+
+class LogisticRegression(LogReg_sklearn):
+    """
+    Sparse Logistic regression scikit-learn estimator based on Celer solver.
+
+    The optimization objective for sparse Logistic regression is::
+
+    mean(log(1 + e^{-y_i x_i^T w})) + 1. / C * ||w||_1
```
Comment (mathurinm) [inline andersoncd/estimators.py:465]: check what we do, sklearn compat or compat with our penalty

File: andersoncd/estimators.py [hunk 16]
```diff
@@ -445,3 +454,201 @@ def path(self, X, y, alphas, coef_init=None, return_n_iter=True, **kwargs):
...
+        self.verbose = verbose
+        self.max_epochs = max_epochs
+        self.p0 = p0
+        self.prune = prune
+        self.max_iter = max_iter
+        self.penalty = penalty
+        self.fit_intercept = fit_intercept
+        self.solver = solver
+        self.datafit = Logistic()
+        self.penalty_ours = L1(1 / C)  # what about the n_samples ?
```
Comment (mathurinm) [inline andersoncd/estimators.py:571]: bad news indeed :(

File: andersoncd/tests/test_estimators.py [hunk 17]
```diff
@@ -45,23 +49,35 @@
 dict_estimators_ours["MCP"] = MCP(
     alpha=alpha, gamma=np.inf, fit_intercept=False, tol=tol)
 
+dict_estimators_sk["LogisticRegression"] = LogReg_sklearn(
+    C=1/(alpha * n_samples), fit_intercept=False, tol=tol, penalty='l1',
+    solver='liblinear', max_iter=100)
```
Comment (mathurinm) [inline andersoncd/tests/test_estimators.py:54]: why specify max_iter ?

File: andersoncd/tests/test_estimators.py [hunk 18]
```diff
@@ -45,23 +49,35 @@
...
 def test_estimator(estimator_name, X):
     # lasso from sklearn
     estimator_sk = dict_estimators_sk[estimator_name]
-    estimator_sk.fit(X, y)
-    coef_sk = estimator_sk.coef_
-
     estimator_ours = dict_estimators_ours[estimator_name]
-    estimator_ours.fit(X, y)
+    if estimator_name == "LogisticRegression":
+        estimator_sk.fit(X, np.sign(y))
```
Comment (mathurinm) [inline andersoncd/tests/test_estimators.py:69]: np.sign can be removed now

File: andersoncd/estimators.py (hunk 11 above)
Comment (QB3) [reply andersoncd/estimators.py:504]: this I have no idea

**Pull request merged**